src/ring_theory/ideal/operations.lean

```diff
@@ -859,6 +859,27 @@ end
 
 end ideal
 
+lemma ideal_mul_eq_bot {R : Type*} [integral_domain R] {I J : ideal R} : (I * J = ⊥) ↔ I = ⊥ ∨ J = ⊥ :=
+begin
+  have hJ : inhabited J, by exact submodule.inhabited J,
+  have j := inhabited.default J, clear hJ,
+  split, swap,
+  { intros,
+    cases a,
+    {rw [← ideal.mul_bot J, a, ideal.mul_comm]},
+    {rw [← ideal.mul_bot I, a, ideal.mul_comm]} },
+  intro hij,
+  by_cases J = ⊥, right, exact h,
+  left,
+  rw submodule.eq_bot_iff,
+  intros i hi,
+  rcases J.ne_bot_iff.1 h with ⟨j', hj, ne0⟩,
+  rw submodule.eq_bot_iff at hij,
+  specialize hij (i * j'),
+  cases eq_zero_or_eq_zero_of_mul_eq_zero (hij $ ideal.mul_mem_mul hi hj),
+  assumption, simpa only [],
+end
```

Comment:
Golfed this proof. Steps:

 * `j` was not used (it will be equal to `0` anyway), so we can get rid of those lines.
 * We don't need to specify `J` in `rw ideal.mul_bot J`, so the two `rw` lines become the same and we can write `cases a; rw`.
 * We can shorten the case distinction on whether `J = ⊥` with the following trick: `refine or_iff_not_imp_left.mpr (λ J_ne_bot, _),`
 * We rewrite `submodule.eq_bot_iff` at the goal and at `hij`, so we can replace it with one `rw at ⊢ hij`.
 * The `cases` where one case is impossible can be replaced with `or.resolve_right`
 * Translate the tactical proof into a term-mode proof dropping into tactic mode when needed

```suggestion
lemma ideal_mul_eq_bot {R : Type*} [integral_domain R] {I J : ideal R} :
  I * J = ⊥ ↔ I = ⊥ ∨ J = ⊥ :=
⟨λ hij, or_iff_not_imp_left.mpr (λ I_ne_bot, J.eq_bot_iff.mpr (λ j hj,
  let ⟨i, hi, ne0⟩ := I.ne_bot_iff.mp I_ne_bot in
    or.resolve_left (mul_eq_zero.mp ((I * J).eq_bot_iff.mp hij _ (mul_mem_mul hi hj))) ne0)),
 λ h, by cases h; rw [← ideal.mul_bot, h, ideal.mul_comm]⟩
```